storage/cleanup.go

```diff
@@ -159,15 +157,60 @@ func (m *cleanupManager) commitLogTimes(t time.Time) (time.Time, []time.Time) {
 	// data loss.
 
 	candidateTimes := timesInRange(earliest, latest, blockSize)
+	namespaces := m.database.GetOwnedNamespaces()
 	cleanupTimes := filterTimes(candidateTimes, func(t time.Time) bool {
-		leftBlockStart, rightBlockStart := t.Add(-blockSize), t.Add(blockSize)
-		anyPendingFlushes := m.fm.NeedsFlush(leftBlockStart, rightBlockStart)
-		return !anyPendingFlushes
+		for _, ns := range namespaces {
+			ropts := ns.Options().RetentionOptions()
+			start, end := commitLogNamespaceBlockTimes(t, blockSize, ropts)
+			if ns.NeedsFlush(start, end) {
+				return false
+			}
+		}
+		return true
 	})
 
 	return earliest, cleanupTimes
 }
 
+// commitLogNamespaceBlockTimes returns the range of namespace block starts which for which the
+// given commit log block may contain data for.
+//
+// consider the situation where we have a single namespace, and a commit log with the following
+// retention options:
+//             buffer past | buffer future | block size
+// namespace:    ns_bp     |     ns_bf     |    ns_bs
+// commit log:     _       |      _        |    cl_bs
+//
+// for the commit log block with start time `t`, we can receive data for a range of namespace
+// blocks depending on the namespace retention options. The range is given by these relationships:
+//	- earliest ns block start = t.Add(-ns_bp).Truncate(ns_bs)
+//  - latest ns block start   = t.Add(cl_bs).Add(ns_bf).Ceil(ns_bs)
+// NB:
+// - blockStart assumed to be aligned to commit log block size
+func commitLogNamespaceBlockTimes(
+	blockStart time.Time,
+	commitlogBlockSize time.Duration,
+	nsRetention retention.Options,
+) (time.Time, time.Time) {
+	earliest := blockStart.Add(-nsRetention.BufferPast()).Truncate(nsRetention.BlockSize())
+	latest := blockStart.Add(commitlogBlockSize).Add(nsRetention.BufferFuture())
+	latest = ceilDate(latest).ceil(nsRetention.BlockSize())
```

Comment:
Should this be `ceil` or `floor`? Seems if `latest` is somewhere in the middle of an ns block, we only need to check if we need to check whether we should flush the current block with a start time of `latest.Truncate(nsBlockSize)` but not the next block?